compiler/rustc_const_eval/src/interpret/eval_context.rs

```diff
@@ -384,7 +385,14 @@ pub(super) fn mir_assign_valid_types<'tcx>(
     // all normal lifetimes are erased, higher-ranked types with their
     // late-bound lifetimes are still around and can lead to type
     // differences.
-    if util::relate_types(tcx, param_env, Variance::Covariant, src.ty, dest.ty) {
+    if util::relate_types(
+        tcx,
+        param_env,
+        DefiningAnchor::Error,
```

Comment:
CTFE is always performed with a `Reveal::All` param-env (or it should, as far as I'm aware?) so it makes sense to use `DefiningAnchor::Error` here.

Reply:
I don't think that's true? We run CTFE for const generics as well and the param-env is different there.
Cc @oli-obk @lcnr

Reply:
My justification is wrong here. We're not using a `Reveal::All` param-env, but we *have* passed this body through the `RevealAll` mir-pass.

Not totally clear if that's *sufficient*, but if it isn't, then we're at least throwing constrained opaques onto the floor during this check in CTFE.

Reply:
We can also approximate the defining anchor from the current stack frame. See newest commit for approach.